src/Features/LanguageServer/Protocol/ExternalAccess/Razor/SemanticTokensRangesHandler.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[Method(SemanticRangesMethodName)]
+internal class SemanticTokensRangesHandler : ILspServiceRequestHandler<SemanticTokensRangesParams, SemanticTokens>
+{
+    private const int TokenSize = 5;
+    public const string SemanticRangesMethodName = "textDocument/semanticTokens/ranges";
+    private readonly SemanticTokensRangeHandler _semanticTokensRangeHandler;
+
+    public SemanticTokensRangesHandler(SemanticTokensRangeHandler semanticTokensRangeHandler)
+    {
+        _semanticTokensRangeHandler = semanticTokensRangeHandler;
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public async Task<SemanticTokens> HandleRequestAsync(
+            SemanticTokensRangesParams request,
+            RequestContext context,
+            CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(request.TextDocument, "TextDocument is null.");
+        if (request.Ranges.Length == 0)
+        {
+            return new SemanticTokens() { Data = Array.Empty<int>() };
+        }
+
+        var responseData = new List<int[]>();
+        foreach (var range in request.Ranges)
+        {
+            var newParams = new SemanticTokensRangeParams
+            {
+                TextDocument = request.TextDocument,
+                Range = range,
+            };
+
+            var semanticTokens = await _semanticTokensRangeHandler.HandleRequestAsync(newParams, context, cancellationToken).ConfigureAwait(false);
```

Comment:
- [ ] I'm gonna explore directly calling `ComputeSemanticTokensDataAsync` rather than injecting the range handler for this call.

https://github.com/dotnet/roslyn/blob/dd7c29de3e172a3abf8768a493d925d54f9a15a9/src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs#L58-L64